inital docker support, closes #42

jhotujec opened the pull request:
Adding initial support for docker containers (#42)

While testing it locally I found out that

- Even though application handles missing `dumps` table, `postgresql` still complains in the logs. Is there any other way to go about this?
- I've had problems accessing the application in Firefox(Firefox Quantum). Did anyone else had this problem?

@johnmuhl

johnmuhl commented:
Thanks for working on this. I have a couple of questions/comments I'll add inline.

The handling for the missing dumps table can absolutely be improved I just haven't got to it yet. If it is causing problems for dockerization I can prioritize fixing it this week.

For Firefox you need to enable ESM support with the `dom.moduleScripts.enabled` setting in `about:config` (look 'ma no transpilers!).

johnmuhl commented on README.md:
```diff
@@ -49,6 +49,10 @@ yarn && yarn dev
 
+## docker
+
+To run a dockerized instances(postgresql and node) for development use `docker-compose up`
```
So this is only meant to make dev work easier? Or could it be used in production instead of the Heroku app? Sorry I'm only superficially familiar with docker.

jhotujec replied:
It's mean for production as well, since `command` parameter can be overwritten at startup with command line. I might add another `docker-compose` just for production though.

johnmuhl commented on docker-compose.yml:
```diff
@@ -0,0 +1,24 @@
+  web:
+    build:
+      context: ./
+      dockerfile: ./server/Dockerfile
```
Is there a reason we don't want to have the dockerfile in the root of the repo?

jhotujec replied:
Whoops, I forgot to move it to the root. Initially I thought I'm gonna have to use two docker files

johnmuhl commented on docker-compose.yml:
```diff
@@ -0,0 +1,24 @@
+      - DATABASE_URL=postgres://example:example@db/electron-crash-report-server
+      - AUTH_USER=crash
+      - AUTH_PASS=electron
+      - PORT=7777
```
Could you use port `40049` here instead. I'm partial to that since it is what I use locally :)

johnmuhl replied:
Or is `7777` a docker idiom?

jhotujec replied:
Ports can be generic. But you wouldn't access it via this port anyway. The ports inside docker containers are being mapped to port `80` on the host. So the way you would access running docker container would be `localhost` with port `80` as default.

johnmuhl replied:
I see now; so ignore that and lets roll with `7777`

johnmuhl replied:
Oops; I see you changed to `8000`. I just mean to roll with whatever port you want and that I don't have a preference here after you explained how it works. `8000`, `7777` or whatever works for me.

jhotujec commented:
@johnmuhl I added a few changes

- Moved `Dockerfile` to root
- We're now using [alpine based docker images](https://hub.docker.com/_/alpine/), which are way smaller
- Shared volumes so that when a file is changed, the data inside container get's automatically refreshed
- Shared volume also for postgres database. Docker containers are ephemeral, so we need a way to persist the data between different docker instances
- Some smaller changes

johnmuhl commented:
I tried to get docker/docker-compose running but ran into a couple of issues. I'll figure those out Monday and test this out but so far it looks good to me. 

I really appreciate you taking this on @jhotujec. With the addition of a production version this will be a huge benefit to those not married to Heroku. Unfortunately I'm not sure how much it will increase source contributions since yours is the first. Maybe it was just the undocumented/janky setup that was holding people back ;-)

Is the dump table thing a blocker for this to land or could we fix that up afterwards?

jhotujec commented:
@johnmuhl no it's not blocking anything.. i just noticed it while testing the application :) What is with the `dom.moduleScripts.enabled` thingy, is this a requirement of one of the frameworks? I'm finishing the production compose file and I'd like to avoid that if possible. This probably has to do wiht imports? Should we compile the code using webpack or something so that it's more robust?

johnmuhl commented:
What should the `submitURL` be when using the docker container? I tried `http://localhost/` but the reports don't show up in the app.

johnmuhl commented:
> What is with the dom.moduleScripts.enabled thingy, is this a requirement of one of the frameworks? I'm
> finishing the production compose file and I'd like to avoid that if possible. This probably has to do with 
> imports? Should we compile the code using webpack or something so that it's more robust?

Firefox is the only browser not supporting javascript modules now so I'd rather just wait for it to catch up than re-add webpack but I'm not going to be religious about it. If it is causing trouble the decision can be reevaluated.

Is there a reason to not have `docker-compose.yml` be the production version and provide a `docker-compose-dev.yml`? Or would it be possible to just have `docker-compose.yml` and use the `NODE_ENV` variable to switch between environments?

jhotujec commented:
I get your point regarding Firefox, but currently market share for Firefox browser is around 13% which is quite a lot. And it looks like it's not working in any Firefox version either(with the addition of Opera Mini and IE11). Does that mean that all of those people aren't gonna be able to see the application?

I'm trying to find the best combination for `docker-compose.production.yml` to support wide(er) array of browser. Maybe we can merge the current version and I will open new PR for production one when I figure it out.

There is no special reason for the namings, but I don't see a problem there. Docker compose file can be selected from command line anyway, meaning that end user can overwrite it if needed. 

I'm also not sure what you mean with `NODE_ENV`, because node is already installed in the container itself. :)

johnmuhl commented:
> I'm also not sure what you mean with `NODE_ENV`

I was asking if it was possible to only have one docker-compose YAML file and use it for both development and production. If not that is OK. Is there anything else you'd like to do here and need me to do before landing this or is it ready as is? I tested it locally and it worked great so I'm happy with the current state if you are.

I will open a new issue to discuss better browser support.

jhotujec commented:
I don't think there are any benefits to trying to merge those two together. I think let's merge this one and I'll open another PR when I'm ready with production compose file, if that's ok :)
